### PR TITLE
Remove logrus from core library

### LINKED ISSUE #39: Abstract sirupsen/logrus for the core CP library
It would be great if logging were configurable, i.e. a custom logger could be injected.

### PR DESCRIPTION
**Pull request by andig**
Fix #39

### CODE REVIEW

**Comment by andig** [comment]
PR compiles now, setting the error handler is still missing. Up for discussion ;)

**Comment by andig** [comment]
ping @lorenzodonini is this going in the direction you want? Anything I could fix?

File: ws/websocket.go [hunk 1]
```diff
@@ -221,19 +237,24 @@ func (server *Server) Start(port int, listenPath string) {
 	server.httpServer.Handler = router
 	if server.tlsCertificatePath != "" && server.tlsCertificateKey != "" {
 		if err := server.httpServer.ListenAndServeTLS(server.tlsCertificatePath, server.tlsCertificateKey); err != http.ErrServerClosed {
-			log.Errorf("websocket server error: %v", err)
+			server.error(fmt.Errorf("failed to listen: %w", err))
```
Comment (lorenzodonini) [inline ws/websocket.go:240]: These statements are failing the CI with `ws/websocket.go:240: Errorf format %w has unknown verb w`. Check the [CI job](https://travis-ci.org/github/lorenzodonini/ocpp-go/jobs/742718154).

CI runs on Go 1.12. Is the `%w` a go 1.13-1.15 feature? If you can confirm this, I will bump the Golang version on CI.

File: ocpp1.6/charge_point.go [hunk 2]
```diff
@@ -247,21 +274,26 @@ func (cp *chargePoint) Start(centralSystemUrl string) error {
 
 func (cp *chargePoint) Stop() {
 	cp.client.Stop()
+
+	if cp.errC != nil {
```
Comment (lorenzodonini) [inline ocpp1.6/charge_point.go:278]: Can you copy the same block also to the Central System's `Stop`?

**Review by lorenzodonini — CHANGES_REQUESTED** [review]
@andig Thanks for the changes, I left a few comments.

Biggest improvement I could see is to use a custom error struct. It's very useful to pass information like `chargePointID` and `FeatureName` as separate fields, so they can be used independently afterwards. What do you think?

File: ws/websocket.go (hunk 1 above)
Comment (andig) [reply ws/websocket.go:240]: Thats 1.13: https://golang.org/doc/go1.13#error_wrapping

File: ocpp1.6/charge_point.go (hunk 2 above)
Comment (andig) [reply ocpp1.6/charge_point.go:278]: Where is that CS stop method?

File: ocpp1.6/central_system.go [hunk 3]
```diff
@@ -371,42 +385,42 @@ func (cs *centralSystem) Start(listenPort int, listenPath string) {
...
 		err := cs.server.SendError(chargePointId, requestId, ocppj.ProtocolError, "Couldn't generate valid confirmation", nil)
 		if err != nil {
-			log.WithFields(log.Fields{
-				"client":  chargePointId,
-				"request": requestId,
-			}).Errorf("unknown error %v while replying to message with CallError", err)
+			err = fmt.Errorf("replying cp %s to request %s with 'protocol error': %w", chargePointId, requestId, err)
+			cs.error(err)
 		}
 	}
```
[the remark this answers is not in the document]
Comment (andig) [reply ocpp1.6/central_system.go:395]: ahh, overlooked.

File: ocpp1.6/charge_point.go (hunk 2 above)
Comment (lorenzodonini) [reply ocpp1.6/charge_point.go:278]: Oh nvm, there is no wrapper function for that yet. I have a ToDo for that, so I'll add it in a later PR.

File: ws/websocket.go (hunk 1 above)
Comment (lorenzodonini) [reply ws/websocket.go:240]: I just updated the base project. Feel free to rebase on master, then the CI should run just fine.

**Comment by andig** [comment]
I've rebased. 

> Biggest improvement I could see is to use a custom error struct. It's very useful to pass information like chargePointID and FeatureName as separate fields, so they can be used independently afterwards. What do you think?

I will add Go 1.13 errors with additional information.

**Comment by lorenzodonini** [comment]
> I've rebased.

Nice!

> 
> > Biggest improvement I could see is to use a custom error struct. It's very useful to pass information like chargePointID and FeatureName as separate fields, so they can be used independently afterwards. What do you think?
> 
> I will add Go 1.13 errors with additional information.

@andig You wanna do this in a separate PR? We could merge this one in the meantime and I'll create a separate issue for the custom error structs.

**Comment by andig** [comment]
Yes, can go. Afaik even structured errors should still be errors so no API break. At least that's what I hope to do...

**Review by lorenzodonini — APPROVED** [review]

**Comment by andig** [comment]
Then lets merge this and I‘ll upgrade the errors in a separate step. Still desperately looking for ws reconnect ;)

**Comment by lorenzodonini** [comment]
@andig merged and created issue #47 as a follow-up. 
Definitely not high priority, but in case you want to help out, I'll appreciate!

Btw, auto-reconnect is working, just verifying network tests and potential race conditions.

**Pull request merged**